core/services/gateway/handlers/functions/orm_test.go

```diff
@@ -0,0 +1,198 @@
+package functions_test
+
+import (
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/functions/generated/functions_router"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/testutils"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/testutils/pgtest"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/handlers/functions"
+)
+
+var (
+	defaultFlags = [32]byte{0x1, 0x2, 0x3}
+)
+
+func setupORM(t *testing.T) (functions.ORM, error) {
+	t.Helper()
+
+	var (
+		db   = pgtest.NewSqlxDB(t)
+		lggr = logger.TestLogger(t)
+	)
+
+	return functions.NewORM(db, lggr, pgtest.NewQConfig(true), testutils.NewAddress())
+}
+
+func createSubscriptions(t *testing.T, orm functions.ORM, amount int) []functions.CachedSubscription {
+	cachedSubscriptions := make([]functions.CachedSubscription, 0)
+	for i := amount; i > 0; i-- {
+		cs := functions.CachedSubscription{
+			SubscriptionID: uint64(i),
+			IFunctionsSubscriptionsSubscription: functions_router.IFunctionsSubscriptionsSubscription{
+				Balance:        assets.Ether(10).ToInt(),
+				Owner:          testutils.NewAddress(),
+				BlockedBalance: assets.Ether(20).ToInt(),
+				ProposedOwner:  common.Address{},
+				Consumers:      []common.Address{},
+				Flags:          defaultFlags,
+			},
+		}
+		cachedSubscriptions = append(cachedSubscriptions, cs)
+		err := orm.UpsertSubscription(cs)
+		require.NoError(t, err)
+	}
+	return cachedSubscriptions
+}
+
+func TestORM_GetSubscriptions(t *testing.T) {
+	t.Parallel()
+	t.Run("fetch first page", func(t *testing.T) {
+		orm, err := setupORM(t)
+		require.NoError(t, err)
+		cachedSubscriptions := createSubscriptions(t, orm, 2)
+		results, err := orm.GetSubscriptions(0, 1)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(results), "incorrect results length")
+		require.Equal(t, results[0], cachedSubscriptions[0])
+	})
+
+	t.Run("fetch second page", func(t *testing.T) {
+		orm, err := setupORM(t)
+		require.NoError(t, err)
+		cachedSubscriptions := createSubscriptions(t, orm, 2)
+		results, err := orm.GetSubscriptions(1, 5)
+		require.NoError(t, err)
+		require.Equal(t, 1, len(results), "incorrect results length")
+		require.Equal(t, results[0], cachedSubscriptions[1])
+	})
+}
+
+func TestORM_UpsertSubscription(t *testing.T) {
```

Comment:
Can you add a test for deleting a subscription? When a sub is cancelled then getSubscriptionsInRange() returns an object with every field empty/zeroed (or does it return nil...?)

Reply:
Checking the contract when a subscription is deleted we are executing delete over the mapping, but Solidity does not delete this element, it just stores empty/default values that subscriptionID. 💡💭  this is why we can rely on order. Will Add a test for this, but it should overwrite a subscription with a balance, etc with an empty balance, etc

Reply:
yeah that's exactly what I wanted to check - if everything works correctly with empty/default values. I see you added a test for that, thanks!